operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+			params.Logger.Info("Starting LB IPAM")
+
+			wg.Add(1)
+
+			// Start go routine, we can't block in OnStart
+			go func() {
+				defer wg.Done()
+				lbIPAM.Run(ipamCtx)
+			}()
+
+			return nil
+		},
+		OnStop: func(ctx hive.HookContext) error {
+			// Close the ipamCtx
+			cancel()
+
+			// Convert blocking Wait into a channel so we can select on the context closing or the Wait.
+			wgDone := make(chan struct{})
+			go func() {
+				wg.Wait()
+				close(wgDone)
+			}()
+
+			select {
+			case <-wgDone:
+			case <-ctx.Done():
+			}
+
+			return nil
+		},
+	})
+
+	return lbIPAM
+}
+
+// LBIPAM is the loadbalancer IP address manager, watcher/controller which allocates and assigns IP addresses
+// to LoadBalancer services from the configured set of LoadBalancerIPPools in the cluster.
+type LBIPAM struct {
+	logger logrus.FieldLogger
+
+	lbClasses   []string
+	ipv4Enabled bool
+	ipv6Enabled bool
+
+	poolClient cilium_client_v2alpha1.CiliumLoadBalancerIPPoolInterface
+	svcClient  client_typed_v1.ServicesGetter
+
+	poolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	svcResource  resource.Resource[*core_v1.Service]
+
+	poolStore    poolStore
+	rangesStore  rangesStore
+	serviceStore serviceStore
+
+	ctx context.Context
+
+	// Only used during testing.
+	initDoneCallbacks []LBIPAMInitDone
+}
+
+type ipPoolEvent = resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+type svcEvent = resource.Event[*core_v1.Service]
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	ipam.ctx = ctx
+
+	errChan := make(chan error, 2)
+
+	poolChan := stream.ToChannel[ipPoolEvent](ctx, errChan, ipam.poolResource)
+	svcChan := stream.ToChannel[svcEvent](ctx, errChan, ipam.svcResource)
+
+	err := ipam.init(ctx, poolChan, svcChan)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while initializing LB-IPAM")
+		// Stop, don't go into main control loop
+		return
+	}
+
+	for _, cb := range ipam.initDoneCallbacks {
+		if cb != nil {
+			cb()
+		}
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are closed to avoid blocking
+	// the resources in case they are re-used after we shutdown.
+	for poolChan != nil && svcChan != nil {
+		select {
+		case event, ok := <-poolChan:
+			if !ok {
+				poolChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.poolOnUpsert,
+				ipam.poolOnDelete,
+			)
+		case event, ok := <-svcChan:
+			if !ok {
+				svcChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.svcOnUpsert,
+				ipam.svcOnDelete,
+			)
+		case err := <-errChan:
+			ipam.logger.WithError(err).Error("Resource error")
+		}
+	}
+}
+
+func (ipam *LBIPAM) init(
+	ctx context.Context,
+	poolChan <-chan ipPoolEvent,
+	svcChan <-chan svcEvent,
+) error {
+	var syncErrors []error
+
+	// First sync all pools
+	poolSync := false
+	for !poolSync {
+		select {
+		case event := <-poolChan:
+			event.Handle(
+				func() error {
+					err := ipam.settleConflicts()
```

Comment:
Shouldn't this only be called once after the store is synced?

Reply:
Nevermind, that's already how it works :-)

Reply:
The first callback is indeed the 'onSync' callback. The resources work in such a way that we first get an upsert for every entry in the resource and then the onSync.

So for example:
* onUpsert pool 1 (event from List)
* onUpsert pool 2 (event from List)
* onSync (Initial sync done)
* onUpsert pool 2 (event from Watch)

We need to call settle conflicts on any change to the pools, not just on sync, the fact that we don't do it during initialization is just a small optimization.